reco_analysis/reco_analysis/chatbot_app/streamlit_app.py

```diff
@@ -284,10 +295,17 @@ def main():
         )
 
     # Authenticate the user
-    authenticator, name, authentication_status, patient_username, patient = setup_authenticator()
+    if (
+        "authentication_status" not in st.session_state
+        or not st.session_state.authentication_status
+    ):
+        authenticator, _, _, _, patient = setup_authenticator()
+        st.session_state.authentication_details = (authenticator, patient)
+    else:
+        authenticator, patient = st.session_state.authentication_details
 
     # Check if the user is authenticated
-    if authentication_status == True and patient is not None:
+    if st.session_state.authentication_status == True and patient is not None:
```

Comment:
this is main change number 2: this prevents `setup_authenticator()` from running with every refresh, which I learned is the main contributor for slowness of each message. This saves about 7 seconds (the bulk of the wait) per doctor message.